### PR TITLE
Update to React 16

### PR DESCRIPTION
**Pull request by nickheiner-usds**
image

[React 16 release notes](https://reactjs.org/blog/2017/09/26/react-v16.0.html).

## Benefits
* We should start eFolder Express fresh on the latest React, instead of writing more code that will need to be migrated eventually. Updating Caseflow to the latest React allows us to share components with eX.
* Various improvements, as described in the release notes. In particular, I'm excited for the ability to return arrays from the `render` functions.
* Reduced file size (ungzipped dev build size is down from `16.8mb` to `13mb`)
* Improved render performance

This PR looks big, but a lot of the diff is just the yarn lockfile.

The test changes are mostly for the new Enzyme API. (We had to update Enzyme to match the latest React.)

### Reader
Some of the legacy Reader unit tests were failing, even though the functionality they were testing worked. We'd see issues like the result of `mount` not being what we would expect, or difficulty with the way we were mocking out API call and dispatching actions. I spent a long time trying to fix the tests, but some were stubborn. I was able to quickly migrate them to rspec, where the test code is much easier to understand, and the test itself exercises more of our app stack. @mdbenjam is ok with this.

### Certification
A bug emerged that was caused by:

1. Overloading the value of redux `state.updateSucceeded` to mean multiple things.
1. Making assumptions about the order in which `componentWillUnmount` and `componentWillMount` are called. In React 15, the behavior could vary by context. In React 16, the behavior is consistent, but it's consistent in a way that breaks our assumption. :smile: From the release notes:
image

Certification used a single state variable, `updateSucceeded`, to indicate whether any server-side update worked. Old pages reset this variable in `componentWillUnmount`, and new pages expected it to be reset in `componentWillMount`. This worked when `oldPage.componentWillUnmount` was always called before `newPage.componentWillMount`, but now it's the other way around. 

I fixed this by introducing a new value to the Redux state that was not overloaded. I tried to do it surgically, so the change in this PR is minimal, but I didn't clean up all cases of the two problems listed above.

### CODE REVIEW

File: client/app/index.js [hunk 1]
```diff
@@ -29,12 +28,6 @@ const COMPONENTS = {
   PerformanceDegradationBanner
 };
 
-// This removes HMR's stupid red error page, which "eats" the errors and
-// you lose valuable information about the line it occurred on from the source map.
-delete AppContainer.prototype.unstable_handleError;
```
Comment (nickheiner-usds) [inline client/app/index.js]: This API is gone, so we don't have to worry about HMR using it poorly any more.

File: client/app/util/initPerfMeasurement.js [hunk 2]
```diff
@@ -1,35 +0,0 @@
-import Perf from 'react-addons-perf';
```
Comment (nickheiner-usds) [inline client/app/util/initPerfMeasurement.js]: `react-addons-perf` does not work in React 16. Instead, the recommendation is to use built-in tools. I've found this to work well locally.

https://reactjs.org/docs/optimizing-performance.html#profiling-components-with-the-chrome-performance-tab

File: client/karma.conf.js [hunk 3]
```diff
@@ -1,6 +1,10 @@
 const _ = require('lodash');
 const webpackConfig = require('./webpack.config.js');
 
+const files = [
```
Comment (nickheiner-usds) [inline client/karma.conf.js]: Just a minor DRYing to save future people from a bug I stumbled in to.

File: client/package.json [hunk 4]
```diff
@@ -45,20 +46,18 @@
     "moment-timezone": "^0.5.13",
     "pdf-annotate.js": "^1.0.0",
     "pdfjs-dist": "1.6.414",
-    "prop-types": "https://registry.npmjs.org/prop-types/-/prop-types-15.5.8.tgz",
```
Comment (nickheiner-usds) [inline client/package.json]: Using the tarball link here was an accident.

File: client/package.json [hunk 5]
```diff
@@ -45,20 +46,18 @@
...
+    "prop-types": "^15.6.0",
     "rc-collapse": "^1.7.6",
-    "react": "^15.5.4",
-    "react-addons-perf": "^15.4.2",
-    "react-copy-to-clipboard": "^4.3.0",
-    "react-dom": "^15.5.4",
+    "react": "^16.2.0",
+    "react-copy-to-clipboard": "^5.0.1",
```
Comment (nickheiner-usds) [inline client/package.json]: I tested this manually and it still works.

File: client/package.json [hunk 6]
```diff
@@ -45,20 +46,18 @@
...
+    "react-dom": "^16.2.0",
     "react-highlight-words": "^0.8.0",
     "react-hot-loader": "^3.0.0-beta.6",
     "react-on-rails": "^8.0.6",
     "react-redux": "^5.0.5",
     "react-router": "^4.1.1",
     "react-router-dom": "^4.1.1",
     "react-select": "^1.0.0-rc.4",
-    "react-test-renderer": "^15.5.4",
```
Comment (nickheiner-usds) [inline client/package.json]: This was not being used afaict.

File: client/package.json [hunk 7]
```diff
@@ -89,7 +88,6 @@
     "karma-sourcemap-loader": "^0.3.7",
     "karma-webpack": "^2.0.3",
     "mocha": "^3.1.2",
-    "react-addons-test-utils": "^15.4.2",
```
Comment (nickheiner-usds) [inline client/package.json]: This was not being used afaict.

File: spec/feature/reader/reader_spec.rb [hunk 8]
```diff
@@ -538,21 +556,6 @@ def expect_doc_type_to_be(doc_type)
       find("#addComment").send_keys(:arrow_right)
       expect_doc_type_to_be "Form 9"
 
-      # Check if annotation mode disappears when moving to another document
-      # Removing this for now. For some reason clicking on the add a comment
```
Comment (nickheiner-usds) [inline spec/feature/reader/reader_spec.rb]: There was no reason to keep this dead code around.

File: client/test/node/reader/DecisionReviewer-test.js [hunk 9]
```diff
@@ -142,177 +145,20 @@ describe('DecisionReviewer', () => {
           expect(wrapper.find('PdfListView')).to.have.length(1);
         }));
       });
-
-      context('rotate', () => {
```
Comment (nickheiner-usds) [inline client/test/node/reader/DecisionReviewer-test.js]: The tests in this file that I wholesale deleted are either already redundant with rspec tests, or I added the necessary rspec tests.

File: client/test/node/reader/DecisionReviewer-test.js [hunk 10]
```diff
@@ -1,5 +1,3 @@
-/* eslint-disable max-lines */
```
Comment (nickheiner-usds) [inline client/test/node/reader/DecisionReviewer-test.js]: image

Comment (nickheiner-usds) [reply client/test/node/reader/DecisionReviewer-test.js]: https://www.youtube.com/watch?v=ZZ5LpwO-An4

File: client/app/certification/reducers/Certification.js [hunk 11]
```diff
@@ -1,9 +1,12 @@
+import { update } from '../../util/ReducerUtil';
+
 export const updateProgressBar = (state, action) => {
   return Object.assign({}, state, {
     currentSection: action.payload.currentSection,
     // reset some parts of state so we don't skip pages or end up in loops
     serverError: null,
     updateSucceeded: null,
+    updateSuccesses: {},
```
Comment (nickheiner-usds) [inline client/app/certification/reducers/Certification.js]: My goal is that eventually, `updateSucceeded` can be removed, and all its functionality can be added as separate entries in `updateSuccesses`.

File: client/app/certification/reducers/index.js [hunk 12]
```diff
@@ -93,6 +93,8 @@ export const certificationReducers = function(state = initialState, action = {})
     return CertificationReducers.handleServerError(state);
   case Constants.CERTIFICATION_UPDATE_SUCCESS:
     return CertificationReducers.certificationUpdateSuccess(state);
+  case Constants.CONFIRM_HEARING_UPDATE_SUCCESS:
+    return CertificationReducers.confirmHearingUpdateSuccess(state);
```
Comment (nickheiner-usds) [inline client/app/certification/reducers/index.js]: This is not the preferred pattern to split up reducers but I certainly was not going to refactor that here. :smile:

File: client/app/reader/CommentIndicator.js [hunk 13]
```diff
@@ -8,7 +8,7 @@ import { ChevronDown, ChevronUp } from '../components/RenderFunctions';
 import Button from '../components/Button';
 import { handleToggleCommentOpened } from '../reader/Documents/DocumentsActions';
 
-class CommentIndicator extends React.PureComponent {
+class CommentIndicator extends React.Component {
```
Comment (nickheiner-usds) [inline client/app/reader/CommentIndicator.js]: This cleans up a warning. In React 16, it complains if you both inherit from `PureComponent` and override `shouldComponentUpdate`.

File: client/test/karma/setup.js [hunk 14]
```diff
@@ -1,12 +0,0 @@
-// eslint-disable-next-line no-empty-function
```
Comment (nickheiner-usds) [inline client/test/karma/setup.js]: Inlining this file with `test-index` simplified the way that Karma and Webpack work together.

File: client/test/node/reader/DecisionReviewer-test.js [hunk 15]
```diff
@@ -104,6 +104,7 @@ describe('DecisionReviewer', () => {
       // the API call to the index endpoint.
         store.dispatch(onReceiveDocs(documents, vacolsId));
         store.dispatch(onReceiveAnnotations(annotations));
+        wrapper.update();
```
Comment (nickheiner-usds) [inline client/test/node/reader/DecisionReviewer-test.js]: For whatever reason, calling this manually is now necessary. #justUnitTestThings

File: client/app/certification/reducers/Certification.js (hunk 11 above)
Comment (mdbenjam) [reply client/app/certification/reducers/Certification.js]: One object where we can look back on all our successes, I like it.

File: client/app/util/initPerfMeasurement.js (hunk 2 above)
Comment (mdbenjam) [reply client/app/util/initPerfMeasurement.js]: Goodbye render police 😢  you will be missed.

image

File: client/test/node/reader/DecisionReviewer-test.js (hunk 10 above)
Comment (mdbenjam) [reply client/test/node/reader/DecisionReviewer-test.js]: Removes a test, then pats himself on the back for bringing the line numbers down. Sounds like something I might do.

File: client/test/helpers/index.js [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+//    <button id="foo">
+//
+// In the example above, we have a React component with an id set as a prop,
+// and a child DOM element with the same id set as an attribute. If we do a
+// simple wrapper.find('#foo'), enzyme will return both elements, and then
+// complain when we call methods that assume only one element was found.
+//
+// To work around this, we manually find the element that's an actual DOM node,
+// and not a React element.
+export const findElementById = (wrapper, id) => wrapper.
```
Comment (mdbenjam) [inline client/test/helpers/index.js]: How did this work before? Is it a bug in the new enzyme?

File: client/test/karma/components/PdfUIPageNumInput-test.js [hunk 17]
```diff
@@ -8,45 +8,46 @@ import { PdfUIPageNumInput } from '../../../app/reader/PdfUIPageNumInput';
...
         docId={1}
-        jumpToPage={() => {
-          return null;
-        }}
+        jumpToPage={() => null}
       />,
       { lifecycleExperimental: true }
     );
 
     it('sets input value correctly', () => {
```
Comment (mdbenjam) [inline client/test/karma/components/PdfUIPageNumInput-test.js]: Were these tests broken? The changes don't seem to differ much. Is the `wrapper.update()` needed with React 16?

Comment (mdbenjam) [reply client/test/karma/components/PdfUIPageNumInput-test.js]: JK saw your comment below. But any idea why it's necessary?

File: client/test/helpers/index.js (hunk 16 above)
Comment (nickheiner-usds) [reply client/test/helpers/index.js]: ¯\\\_(ツ)\_/¯

**Comment by amprokop** [comment]
@NickHeiner — can you explain the mount/unmount Certification bug?

File: client/app/certification/reducers/Certification.js (hunk 11 above)
Comment (nickheiner-usds) [reply client/app/certification/reducers/Certification.js]: #memories

**Pull request merged**